File: 29Items-Projects/7Algorithms/4-Greedy-algorithms/java/src/main/java/com/example/greedy/algorithms/ActivitySelection.java
package com.example.greedy.algorithms;

import com.example.greedy.model.Activity;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public final class ActivitySelection {
    private ActivitySelection() {
    }

    public static List<Activity> select(List<Activity> activities) {
        for (Activity activity : activities) {
            if (activity.finish() < activity.start()) {
                throw new IllegalArgumentException(activity.name() + " finishes before it starts");
            }
        }

        List<Activity> sorted = activities.stream()
                .sorted(Comparator.comparingInt(Activity::finish)
                        .thenComparingInt(Activity::start)
                        .thenComparing(Activity::name))
                .collect(Collectors.toList());

        List<Activity> selected = new ArrayList<>();
        int currentFinish = -1;
        for (Activity activity : sorted) {
            if (activity.start() >= currentFinish) {
                selected.add(activity);
                currentFinish = activity.finish();
            }
        }
        return selected;
    }
}

File: 29Items-Projects/7Algorithms/4-Greedy-algorithms/python/greedy_algorithms/algorithms/activity_selection.py
from dataclasses import dataclass


@dataclass(frozen=True)
class Activity:
    name: str
    start: int
    finish: int


def select_activities(activities: list[Activity]) -> list[Activity]:
    for activity in activities:
        if activity.finish < activity.start:
            raise ValueError(f"{activity.name} finishes before it starts")

    selected: list[Activity] = []
    current_finish = -1
    for activity in sorted(activities, key=lambda item: (item.finish, item.start, item.name)):
        if activity.start >= current_finish:
            selected.append(activity)
            current_finish = activity.finish
    return selected
